### PR TITLE
Optimise animations a bit

### PR DESCRIPTION
**Pull request by metalgearsloth**
Also did some ECS at the time.

### CODE REVIEW

File: Robust.Client/GameObjects/EntitySystems/AnimationPlayerSystem.cs [hunk 1]
```diff
@@ -1,15 +1,64 @@
...
-                animationPlayerComponent.Update(frameTime);
+                var anim = _activeAnimations[i];
+                if (!Update(anim, frameTime)) continue;
+                _activeAnimations.RemoveAt(i);
             }
         }
+
+        internal void AddComponent(AnimationPlayerComponent component)
+        {
+            if (_activeAnimations.Contains(component)) return;
```
Comment (PJB3005) [inline Robust.Client/GameObjects/EntitySystems/AnimationPlayerSystem.cs]: This being a list could be quite problematic perf wise if many animation players start playing at once. Maybe keep a bool flag on the component "I am currently playing" instead?

File: Robust.Client/GameObjects/EntitySystems/AnimationPlayerSystem.cs [hunk 2]
```diff
@@ -1,15 +1,64 @@
...
+
         public override void FrameUpdate(float frameTime)
         {
-            foreach (var animationPlayerComponent in EntityManager.ComponentManager.EntityQuery<AnimationPlayerComponent>(true))
+            for (var i = _activeAnimations.Count - 1; i >= 0; i--)
             {
-                animationPlayerComponent.Update(frameTime);
+                var anim = _activeAnimations[i];
+                if (!Update(anim, frameTime)) continue;
+                _activeAnimations.RemoveAt(i);
```
Comment (PJB3005) [inline Robust.Client/GameObjects/EntitySystems/AnimationPlayerSystem.cs]: This is again really bad perf wise. Use `RemoveSwap()` instead?

File: Robust.Client/GameObjects/EntitySystems/AnimationPlayerSystem.cs [hunk 3]
```diff
@@ -1,15 +1,64 @@
...
+            if (_activeAnimations.Contains(component)) return;
+            _activeAnimations.Add(component);
+        }
+
+        private bool Update(AnimationPlayerComponent component, float frameTime)
+        {
+            if (component.PlayingAnimationCount == 0)
+                return true;
+
+            List<string>? toRemove = null;
```
Comment (PJB3005) [inline Robust.Client/GameObjects/EntitySystems/AnimationPlayerSystem.cs]: You can use `RemQueue<T>` for this now.

**Comment by PJB3005** [comment]
This causes a very gnarly test failure related to dummy icons by the way. Not sure how to go about handling that.

File: Robust.Client/GameObjects/EntitySystems/AnimationPlayerSystem.cs (hunk 1 above)
Comment (metalgearsloth) [reply Robust.Client/GameObjects/EntitySystems/AnimationPlayerSystem.cs]: good idea

File: Robust.Client/GameObjects/EntitySystems/AnimationPlayerSystem.cs (hunk 3 above)
Comment (metalgearsloth) [reply Robust.Client/GameObjects/EntitySystems/AnimationPlayerSystem.cs]: Unfortunately I did consider this but the greatness that is RemQueue is on content (for now)

Comment (metalgearsloth) [reply Robust.Client/GameObjects/EntitySystems/AnimationPlayerSystem.cs]: Alright made a new PR for https://github.com/space-wizards/RobustToolbox/pull/2066 so will update after that's merged

**Pull request merged**